File: algorithms/092.Reverse_Linked_List_II/Reverse_Linked_List_II.py
<<<<<<< HEAD
# Time:  O(n)
# Space: O(1)

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
    
    def __repr__(self):
        if self:
            return "{} -> {}".format(self.val, repr(self.next))
class Solution:
    # @param head, a ListNode
    # @param m, an integer
    # @param n, an integer
    # @return a ListNode
    def reverseBetween(self, head, m, n):
        diff, dummy, cur = n - m + 1, ListNode(-1), head
        dummy.next = head
        
        last_unswapped = dummy
        while cur and m > 1:
            cur, last_unswapped, m = cur.next, cur, m - 1
            
        prev, first_swapped = last_unswapped,  cur
        while cur and diff > 0:
            cur.next, prev, cur, diff = prev, cur, cur.next, diff - 1
            
        last_unswapped.next, first_swapped.next = prev, cur
        
        return dummy.next

if __name__ == "__main__":
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    print Solution().reverseBetween(head, 2, 4)
            

=======
# Time:  O(n)
# Space: O(1)

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
    
    def __repr__(self):
        if self:
            return "{} -> {}".format(self.val, repr(self.next))
class Solution:
    # @param head, a ListNode
    # @param m, an integer
    # @param n, an integer
    # @return a ListNode
    def reverseBetween(self, head, m, n):
        diff, dummy, cur = n - m + 1, ListNode(-1), head
        dummy.next = head
        
        last_unswapped = dummy
        while cur and m > 1:
            cur, last_unswapped, m = cur.next, cur, m - 1
            
        prev, first_swapped = last_unswapped,  cur
        while cur and diff > 0:
            cur.next, prev, cur, diff = prev, cur, cur.next, diff - 1
            
        last_unswapped.next, first_swapped.next = prev, cur
        
        return dummy.next

if __name__ == "__main__":
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    print Solution().reverseBetween(head, 2, 4)
            

>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74

File: algorithms/092.Reverse_Linked_List_II/Reverse_Linked_List_II.cpp
#include <bits/stdc++.h>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        ListNode *mhead = new ListNode(0), *prev, *cur;
        mhead->next = head;    // because m will be 0
        for (int i = 0; i < m - 1; i++) {
            mhead = mhead->next;
        }

        prev = mhead->next;
        cur = prev->next;
        for (int i = m; i < n; i++) {
            prev->next = cur->next;
            cur->next = mhead->next;
            mhead->next = cur;
            cur = prev->next;
        }
        return m == 1 ? mhead->next : head;
    }
};
class Solution2 {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode **p=&head;
        for(int i=1;i<m;i++)
            p=&(*p)->next;
        ListNode **pos=&(*p);
        for(int i=m;i<n;i++){
            ListNode *t=(*p)->next;
            (*p)->next=t->next;
            t->next=*pos;
            *pos=t;
            if(i==m)
                p=&(*p)->next;
        }
        return head;
    }
};
class Solution3 {
public:
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        //make a head by myself
        ListNode *newHead=new ListNode(0);
        newHead->next=head;
        ListNode *p0=newHead;
        ListNode *p1=nullptr,*p2=nullptr;
        int i;
        //find position of p0
        for(i=1;i<m;i++)
            p0=p0->next;
        p1=p0->next;
        //reverse one by one
        while(i<n)
        {
            p2=p1->next;
            p1->next=p2->next;
            p2->next=p0->next;
            p0->next=p2;
            i++;
        }

        // p is the true head

        ListNode *p=newHead->next;
        delete newHead;
        return p;
    }
};

int main() {

    return 0;
}

